Cosmic text 0.11

bytemunch opened the pull request:
Updates `cosmic-text` to 0.11, rips out inbuilt functionality in favour of a more modular design.

Todo before merging:
- [x] Clicking in a text box that has been edited without being unfocused places the cursor incorrectly
- [x] Fix setting a `FocusedWidget` before widget instantiation (i.e in `Startup`)
- [x] Configurable cursor, selection colors
- [x] Run layout and shaping systems only when needed ~~(And whenever needed, dynamically setting text doesn't update)~~
- [x] Update examples with new API
- [x] Pass tests

Embrace regression, create modularity
~~- [ ] Reimplement placeholders~~
~~- [ ] Reimplement password fields~~
~~- [ ] Reimplement undo/redo~~
~~- [ ] Check for other regressions~~

bytemunch committed:
barely got text entry working again

bytemunch committed:
remove local optimizations

i should really put them somewhere else on my machine

Dimchikkk commented on src/lib.rs:
```diff
@@ -103,31 +103,29 @@ pub struct ReadOnly; // tag component
-#[derive(Component)]
-pub struct CosmicEditor(pub Editor);
+#[derive(Component, Deref, DerefMut)]
+pub struct CosmicEditor(pub Editor<'static>);
```
👍

bytemunch committed:
delete most stuff to get the basics nearly working

bytemunch commented:
Example `basic_ui` is where I'm testing these changes, the widget works for editing when defocused (ESC) and then focused again. Still doesn't modify the underlying buffer, borrow checker skill issues lol

StaffEngineer commented:
Let me try to test `basic_ui` today ;)

StaffEngineer commented:
It's white screen for me :) but I see some widget text in console

bytemunch committed:
editable text

lots of cloning but seems to be the only way if buffers are in the ECS

editor widgets don't relayout when edited, but when unfocused the buffer does

StaffEngineer commented:
editable banana, nice :D

Commit:
fix backspace on native and make basic_ui wasm editable

bytemunch committed:
fix mid-edit layout

bytemunch committed:
reimplement text cursor blinking

uhoh it's looking like im using the not a starting point pr as a starting point lol

StaffEngineer commented:
@bytemunch we need to have some plan how bring this branch into main... what the scope of tasks we have to deliver to make it work, any ideas? :)

bytemunch commented:
Updated the PR message with a checklist (from the top of my head, I'm sure there'll be more to get done but so far unknown unknowns)

Hopefully most of the work can be easily reimported from the main branch, but with so many code changes already compatibility is uncertain.

Feel free to edit the checklist with anything you can think of that I have missed, lmk if this scope sounds appropriate

StaffEngineer commented:
What comes to mind... 

would be cool to have `core` module that supports only essential things: like input handling and drawing text and then move everything else (like password, placeholder, auto-height, undo-redo, etc) into separate modules (bevy plugins maybe?). Undo-redo wasn't implemented perfectly for example... there is https://github.com/pop-os/cosmic_undo_2 that we could take inspiration from... if we split undo-redo functionality to another plugin it would be easier to refactor/improve it in this case.

another thing - ab_glyph that used for text rendering in bevy can give position of "text curves" to draw text in 3d... if cosmic-text can give us that.. and we can't marry vello + cosmic-text + bevy to use GPU for text rendering we could think about how we can speed text drawing up... maybe to use rayon to utilise all CPUs for pixel calculation.

bytemunch commented:
Yeah that sounds like a more maintainable way forward, though I do find the bevy plugin dependency issue pretty difficult to work with. Maybe for now we could use internal plugins in the same crate (lose for binary size but makes plugin interop a non issue), and once https://github.com/bevyengine/bevy/issues/69 is fixed split modules out to their own crates?

Binary size issue can be addressed with feature flags for toggling internal plugins on and off.

Dropping scope to a core module is a great move though, unix philosophy and all that.

StaffEngineer commented:
Yep, internal plugins sounds good.

bytemunch committed:
update `basic_sprite` example

adds a util crate for common example systems
also removes history tracking structs

bytemunch committed:
`image_background`+`bevy_api_testing` examples

MV `bevy_api_testing.rs` > `sprite_and_ui_clickable.rs`

bytemunch committed:
remove `CosmicText` in favor of buffer functions

simplifys API and mirrors Bevy's Text functions more closely

bytemunch committed:
strip autoheight + example, update other examples

StaffEngineer commented:
so close no matter how far :)

bytemunch committed:
fix wasm compilation

bytemunch committed:
clippy

Commit:
fix text

Commit:
fix clippy

Dimchikkk committed:
fix wasm

Dimchikkk committed:
fix import

Dimchikkk committed:
add changelog

bytemunch committed:
Merge branch 'testing_cosmic-text_0.11' of github.com:StaffEngineer/bevy_cosmic_edit into testing_cosmic-text_0.11

bytemunch commented:
Aah i have duplicate commits from working offline :laughing: Just pushed cos I couldn't figure out how to marry them >.<

bytemunch committed:
fix mouse cursor hover on sprite widget

bytemunch committed:
fix editor cursor location when edited

StaffEngineer commented:
@bytemunch could you elaborate on "Fix setting a FocusedWidget"?

bytemunch committed:
fix blank widget when focused on startup

bytemunch commented:
Got it :) was letting `set_widget_size` and `set_buffer_size` run unordered, when they need to be in order

StaffEngineer commented:
Makes sense.... good job 👍 

Just throwing some info at you :D

I made a small check yesterday: it's indeed possible to get outline curves using `cosmic-text` for the given font... and I know that it's possible to create meshes for each glyph using `lyon`... I am brainstorming whether it makes sense to adopt this approach in the future... instead of generating pixels using cosmic-text draw calls and replacing bevy asset... do something like this: https://github.com/tigregalis/bevy_text3d but using `cosmic-text` for obtaining outline commands to draw text.

bytemunch commented:
Definitely something to look into, could be better for both perf and rendering. 

I'll throw some ideas back lol.

If we can somehow make the rendering itself an (internal) plugin then creating a mesh renderer would be simpler. Figuring out how the upcoming internal plugin architecture works should show if and how such a system can work. I'll try implementing placeholders and password fields as plugins after this PR meets goals, that should set the groundwork for a pluggable renderer.
Rambling a bit but I guess I'm saying decoupling rendering and data allows for everything above and more. Plus, a data-only core module could just slot straight into bevy upstream, far as I understand the inner bevy components are essentially plugins themselves.

bytemunch committed:
configurable cursor/selection colors

bytemunch committed:
run layout functions when needed

bytemunch committed:
fix ui widget mouse cursor

StaffEngineer commented:
Yeah, abstracting renderer sounds cool 👍 

Then, `bevy_cosmit_edit` can have multiple renderers...

Dimchikkk committed:
fix scale change

bytemunch commented:
> There is tiny bug: when widget becomes active cursor first jumps to the beginning of buffer then on desired position but I think we can live with that

Yeah I noticed, will have to time it to find out more, if it's a single frame delay then it could be a systems ordering thing
